marconi/src/Marconi/IndexersHotStore.hs

```diff
@@ -0,0 +1,172 @@
+update hotStore k v =
+    let
+        insertValues  = KeyValues .  Map.adjust (Set.insert v) k  . unKeyValues
+    in
+         (atomically . modifyTVar (unHot hotStore)) insertValues
+
+-- | returs a sorted list of keys
+keys :: HotStore k v -> IO (Set k)
+keys hotStore = pure . Map.keysSet . unKeyValues =<< (readTVarIO . unHot) hotStore
+
+findKey :: Ord k => HotStore k v -> k -> IO (Set v)
+findKey hotStore k =  do
+    m <- readTVarIO . unHot $ hotStore
+    case Map.lookup k . unKeyValues $ m  of
+        Just vs -> pure vs
+        _       -> pure Set.empty
+
+-- | Retrieve a Set of TxOutRefs associated with the given plutus address
+-- We return an empty Set if no address is found
+findByPlutusAddress
+    :: IndexerHotStore      -- ^ hot store
+    -> Address              -- ^ Plutus address
+    -> IO (Set TxOutRef)    -- ^ set of corresponding TxOutRefs
+findByPlutusAddress  =  findKey
+
+-- | Retrieve a Set of TxOutRefs associated with the given Cardano Era address
+-- We return an empty Set if no address is found
+findByCardanoAddress
+    :: IndexerHotStore                              -- ^ hot store
+    -> CApi.Address addrtype                        -- ^ Cardano address and Era
+    -> IO (Either ToCardanoError (Set TxOutRef) )   -- ^ To Plutus address conversion error may occure
+findByCardanoAddress hotStore address =
+     case fromCardanoAddress address of
+         Left e              -> pure . Left $ Tag (show e) DeserialisationError
+         Right plutusAddress -> Right <$> findByPlutusAddress hotStore plutusAddress
+
+-- | Retrieve a Set of TxOutRefs associated with the given Cardano Era address
+-- We return an empty Set if no address is found
+findByAddress
+    :: IndexerHotStore                              -- ^ hot store
+    -> Text                                         -- ^ Bech32 Address
+    -> IO (Either ToCardanoError (Set TxOutRef) )   -- ^ To Plutus address conversion error may occure
+findByAddress hotStore addressText =
+    let
+        addressEither :: Either CApi.Bech32DecodeError  CardanoAddress
+        addressEither = CApi.deserialiseFromBech32 (CApi.proxyToAsType Proxy) addressText
+    in
+        case addressEither of
+            Right address -> findByCardanoAddress hotStore address
+            Left _        -> pure . Left $ Tag (unpack  addressText) DeserialisationError
+
+findAll :: HotStore k v -> IO [(k,Set v)]
+findAll hotStore = Map.toList . unKeyValues <$> (readTVarIO . unHot $ hotStore)
+
+type CardanoAddress = CApi.Address CApi.ShelleyAddr
+
+-- | Typre represents non empty list of Bech32 compatable addresses"
+type TargetAddresses = NonEmpty CardanoAddress
+
+-- | does the transaction contain a targetAddress
+isInTargetTxOut
+    :: TargetAddresses
+    -- ^ non empty list of target address
+    -> CApi.TxOut CApi.CtxTx era
+    -- ^  a cardano transaction out that contains an address
+    -> Bool
+isInTargetTxOut targetAddresses (CApi.TxOut address _ _) = case  address of
```

Comment:
@raduom , Marconi.Indexer will use this predicate to filter for addresses of interest to the users